airflow/providers/cncf/kubernetes/utils/pod_launcher.py

```diff
@@ -117,79 +129,99 @@ def delete_pod(self, pod: V1Pod) -> None:
-                delta = dt.now() - curr_time
-                if delta.total_seconds() >= startup_timeout:
-                    msg = (
-                        f"Pod took longer than {startup_timeout} seconds to start. "
-                        "Check the pod events in kubernetes to determine why."
-                    )
-                    raise AirflowException(msg)
-                time.sleep(1)
-
-    def monitor_pod(self, pod: V1Pod, get_logs: bool) -> Tuple[State, V1Pod, Optional[str]]:
+        while True:
+            remote_pod = self.read_pod(pod)
+            if remote_pod.status.phase != PodStatus.PENDING:
+                break
+            self.log.warning("Pod not yet started: %s", pod.metadata.name)
+            delta = dt.now() - curr_time
+            if delta.total_seconds() >= startup_timeout:
+                msg = (
+                    f"Pod took longer than {startup_timeout} seconds to start. "
+                    "Check the pod events in kubernetes to determine why."
+                )
+                raise PodLaunchFailedException(msg)
+            time.sleep(1)
+
+    def follow_container_logs(self, pod: V1Pod, container_name: str):
+        """
+        Follows the logs of container and streams to airflow logging.
+        Returns when container exits.
+        """
+        container_stopped = False
+        read_logs_since_sec = None
+        last_log_time = None
+
+        # `read_pod_logs` follows the logs so we shouldn't necessarily _need_ to loop
+        # but in a long-running process we might lose connectivity and this way we
+        # can resume following the logs
+        while True:
+            try:
+                logs = self.read_pod_logs(
+                    pod=pod,
+                    container_name=container_name,
+                    timestamps=True,
+                    since_seconds=read_logs_since_sec,
+                )
+                for line in logs:  # type: bytes
+                    timestamp, message = self.parse_log_line(line.decode('utf-8'))
+                    self.log.info(message)
+                    if timestamp:
+                        last_log_time = timestamp
+            except BaseHTTPError:  # Catches errors like ProtocolError(TimeoutError).
+                self.log.warning(
+                    'Failed to read logs for pod %s',
+                    pod.metadata.name,
+                    exc_info=True,
+                )
+
+            if container_stopped is True:
+                break
+
+            if last_log_time:
+                delta = pendulum.now() - last_log_time
+                read_logs_since_sec = math.ceil(delta.total_seconds())
+
+            time.sleep(1)
+
+            if self.container_is_running(pod, container_name=container_name):
+                self.log.info('Container %s is running', pod.metadata.name)
+                self.log.warning('Pod %s log read interrupted', pod.metadata.name)
```

Comment:
Under normal circumstances, we'd only stop getting log lines when the container has stopped, so these lines happen when something sorta abnormal has happened (most likely something that was a `BaseHTTPError`).

Reply:
i will look at this method to see if i can help clarify the intention  / behavior overall cus it was not obvious to me either why it was looping

Reply:
ok i did refactor this method for better readability

but i also took another look and saw that in the current we're a little less defensive (i was reading  one  more loop in case container completed in between the log following and the container status read).

i changed it back to  exit immediately if container is stopped (which makes the code simpler) but i moved  the  sleep down a bit to further reduce likelihood of log loss.

Reply:
I like the refactor, however, that "read once more on failure" behavior does defend against a race condition (though unlikely):

1. log retrieval failure (BaseHTTPError)
2. container logs more
3. finishes
4. then we check the status, it's stopped 
5. we short-circuit out, missing the last logs

Reply:
We had a chat and we are going to leave this unlikely race condition alone for now.